pallets/parachain-system/src/validate_block/timestamp.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus. If not, see <http://www.gnu.org/licenses/>.
+
+//! Utility module to validate cumulus-runtime timestamps against the
+//! relay chain slot.
+
+use cumulus_primitives_core::{OnValidationData, Slot};
```

Comment:
This should be its own pallet. Similar to `aura-ext`. Maybe `timestamp-ext` or something.

The idea with using environmental was also a bad idea by me. We should probably store the intermediate data in the storage.